README
This is version of evtouch driver with 4-point calibration, based on version from 
http://linux.chapter7.ch/touchkit (by Daniel Ritz).

Have Fun
    Gennady


ChangeLog
0.8.7-HR *** Note this is not an official release but an "unauthorized
	 *** fix version by me (Harald Radke - harryrat@postnuklear.de)
	-> Changed code to compile a working driver for  both 
	  XInput ABI 0.x and 2.0 (Xorg 1.4.0)
	  (test compile for 0.x still needs to be done)
	  - evtouch.c: 
            * capsulated xf86Post*Events and xf86InputSetScreen 
	      in functions where depending on the XINPUT ABI conversion
	      is done or xf86 functions are directly called
	    * xf86InitValuatorAxisStruct and xf86AlwaysCore() depending
	      on ABI done
	  - libtouch.h: passing xf86PostButtonEvent capsulation function
	    to init function (same as in evtouch.c) to do proper button
	    triggering
	-> Rotation:
	  - added support for Option "Rotate" "UD" (upside down) 
	     in evtouch input section of xorg.conf
	  - due to Segfault with (at least) fbdev video device, only
	    evtouch "Rotate" option is taken into calculations for coords,
	    meaning to get proper touchscreen behaviour on rotated X screens.
	    same Rotate option as in video device section of xorg.conf has
            to be set in the evtouch section
	-> Drag support:
	  Option "<STATE>_drag" "<BUTTON_ID>" has been added. When set,
	  the handle_<STATE>() function triggers BUTTON_DOWN for Button <BUTTON_ID>
	  before changing into S_MOVE state. Note that actually only
	  Option "touched_drag" "<BUTTON_ID>" is implemented, rest is dummy
	  (and doesn't make too much sense I guess) (-;
	-> tap "tolerance":
	   introduce move_limit to evtouch.c as it is in libtouch.c, using it
	   to discard ABS_X/Y position changes if they are inside move_limit
	   around cur_x/y. this way pointer hopping can be avoided as well as
           proper touched_drag established
         - changes of cur_x/y are reported on EV_SYN in evtouch.c ReadInput() instead
           on EV_ABS
	-> added generic ts-adc touchscreen device to udev rules file	   

	=>  Example for a Loox720 PDA setup in input section of xorg.conf:
		Section "InputDevice"
		 Identifier "touchscreen"
		 Driver "evtouch"
		 Option "Device" "/dev/input/evtouch_event"
		 Option "MinX" "434"
		 Option "MinY" "333"
		 Option "MaxX" "3643"
		 Option "MaxY" "3804"
		 Option "SwapX" "true"
		 Option "longtouched_action" "down"
		 Option "longtouched_button" "3"
		 Option "maybetapped_action" "click"
		 Option "maybetapped_button" "1"
		 Option "touched_drag" "1"
		 Option "oneandahalftap_button" "0"
		 Option "TapTimer" "30"
		 Option "LongtouchTimer"  "500"
		 Option "MoveLimit"  "18"
		EndSection

0.8.7
	- Added Elo Touchscreen to udev-rules (Stephen Webb)
	- Added "fix" for AMD64-crashes on touch (Martin Letenay)
	- Fixed driver-behavior when resolution is changed using xrandr
	  (Yura Semashko)
	- Added stuff for installing udev-rules to /etc/... (Tuuka
	  Pasanen)
	- Added workaround for kernel-driver reporting ABS_Z and ABS_RX
	  instead of ABS_X and ABS_Y (Lutz Sammer + me)

0.8.6
	- replaced old udev-rules with a more complete one (Thomas Bartosik)
	- Added Lifebook B-series to udev-rules
	- Fixed classic rotation (again)
	- Fixed wrong handling of wrong / fixed default screen size
	- little code-cleanup (removed obsolete variables)
	- Fixed segv when switching to VT (Mattia Dongili)

0.8.5
	- Made Events in libtouch configurable (experimental)
	- Changed initialization for Xorg 7.2

0.8.4
	- Added XRandr-support (Mattia Dongili)
	- Fixed Error in calibrate.sh
	- Minor Compile-Warning fixes (Tuukka Pasanen)
	- Make CURSORDIR work again (Tuukka Pasanen)
	- added udev-rules for touchscreen device detection (Tuuka
	  Pasanen)
	- Button-press / -release now are alway issued on EV_SYN (me)

0.8.3:
	- removed useless trigger_sm
	- fixed Middle-Button-Handling again

0.8.2:
	- Autotools-Improvements by Mattia Dongili
	- pen-touch logic is used from libtouch now
	- Minor improvements for libtouch (for handling evtouch)
	- Middle button and Scrollwheel-Support (by Anthony Liguori)

0.8.1:
	- Applied autotools-patch from Tuukka Pasanen

0.8:
	- Imake-Cleanup
	- Switched default-Y-direction (SDL + GPM compatibility)
	- Implementation of Statemachine for the pen
	  - Configuration changed (see Website for Details)
	  - Behaviour for the pen has changed

0.6.1:
	- Added support for swapping y-axis
	- Added support for swapping x-axis
	- Added some missing files for calibration 
	  (forgot to add them to the tarball)
	- Fixed bug/typo in CW-rotation 

	(Thanks to Jasun Wurster who
	provided useful input and testing on the M150 and M170
	touchsreens for the 0.6.1 release)

0.6:
	- Added support for screen-rotation (Thanks to Kevin Quigley
	for some very useful input) 
	- Added support for calibration

0.5.1:
	- Added support for emulate3Buttons

0.5: 
	- Initial Release


calibrator.c
/*
 * simple and generic touchscreen calibration program using the linux 2.6
 * input event interface
 *
 */

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <X11/Xatom.h>
#include <X11/keysym.h>

#include <stdio.h>
#include <signal.h>
#include <termios.h>
#include <stdlib.h>
#include <sys/time.h>

/*****************************************************************************/
/* stuff for event interface */
struct input_event {
	struct timeval time;
	unsigned short type;
	unsigned short code;
	int value;
};

/* event types */
#define EV_SYN			0x00
#define EV_KEY			0x01
#define EV_REL			0x02
#define EV_ABS			0x03

/* codes */
#define ABS_X			0x00
#define ABS_Y			0x01
#define SYN_REPORT		0
#define BTN_LEFT		0x110
#define BTN_RIGHT		0x111
#define BTN_TOUCH		0x14a

/*****************************************************************************/
/* some constants */
#define FONT_NAME		"9x15"
#define IDLETIMEOUT		15
#define BLINKPERD		0.16        

#define ROUND_SYMBOL
#define NumRect			5

#if 0 
#define Background		cCYAN
#define	TouchedCross		cYELLOW
#define BlinkCrossBG		cRED
#define BlinkCrossFG		cWHITE
#define nTouchedCross		cBLUE
#define Cross			cWHITE
#define DrawGrid		cWHITE
#define DrawLine		cYELLOW
#define DrawUp			cRED
#define DrawDown		cBLUE
#define TimerLine		cRED
#define PromptText		cBLUE
#else
#define Background		cBLACK
#define	TouchedCross		cYELLOW
#define BlinkCrossBG		cRED
#define BlinkCrossFG		cWHITE
#define nTouchedCross		cBLUE
#define Cross			cYELLOW
#define DrawGrid		cGREEN
#define DrawLine		cYELLOW
#define DrawUp			cRED
#define DrawDown		cGREEN
#define TimerLine		cRED
#define PromptText		cWHITE
#endif

#define N_Colors		10

static char colors[N_Colors][10] =
{ "BLACK", "WHITE", "RED", "YELLOW", "GREEN", "BLUE", "#40C0C0" };
				
static unsigned long pixels[N_Colors];

#define cBLACK			(pixels[0])
#define cWHITE			(pixels[1])
#define cRED			(pixels[2])
#define cYELLOW			(pixels[3])
#define cGREEN			(pixels[4])
#define cBLUE			(pixels[5])
#define cCYAN			(pixels[6])

/* some stupid loops */
#define SYS_1( zzz... ) do {	\
	while ( (zzz) != 1 );	\
} while (0)

#define SYS_0( zzz... ) do {	\
	while ( (zzz) != 0 );	\
} while (0)

/* where the calibration points are placed */
#define SCREEN_DIVIDE	16
#define SCREEN_MAX	0x800
#define M_POINT		(SCREEN_MAX/SCREEN_DIVIDE)
int MARK_POINT[] = { M_POINT, SCREEN_MAX - 1 - M_POINT };

/*****************************************************************************/
/* globals */
int job_done = 0;
int points_touched = 0;
char *deviceName;
int points_x[4], points_y[4];

Display *display;
int screen;
GC gc;
Window root;
Window win;
XFontStruct *font_info;
unsigned int width, height;	/* window size */
char *progname;
int evfd;


/*****************************************************************************/

int x = -1, y = -1;

int get_events(int *px, int *py)
{
	int ret;
	int touch = 0, sync = 0;
	struct input_event ev;
    int limit = 10;

	/* read till sync event */
	while (sync == 0 || touch == 0) {
		ret = read(evfd, &ev, sizeof(ev));
		if (ret == -1)
			return -1;

		switch (ev.type) {
		case EV_ABS:
			switch (ev.code) {
			case ABS_X:
				if (x == -1)
					x = ev.value;
				break;

			case ABS_Y:
				if (y == -1)
					y = ev.value;
				break;

			default:
                printf("Unknown ev.code=%d for ev.type=EV_ABS\n", ev.code);
				break;
			}

			break;

		case EV_KEY:
			switch (ev.code) {
			case BTN_LEFT:
			case BTN_TOUCH:
				touch = 1;
                break;

			default:
                printf("Unknown ev.code=%d for ev.type=EV_KEY\n", ev.code);
			    break;
			}

			break;

		case EV_SYN:
			if (ev.code == SYN_REPORT)
				sync = 1;
            else
                printf("Unknown ev.code=%d for ev.type=EV_SYN\n", ev.code);


			break;

		default:
            printf("Unknown ev.type=EV_SYN\n", ev.type);
			break;
		}
	}

	if (!touch || x == -1 || y == -1)
		return -1;

	*px = x;
	*py = y;

    x = -1;
    y = -1;

	return 0;
}


/*****************************************************************************/

void cleanup_exit()
{
	SYS_1(XUnloadFont(display, font_info->fid));
	XUngrabServer(display);
	XUngrabKeyboard(display, CurrentTime);
	SYS_1(XFreeGC(display, gc));
	SYS_0(XCloseDisplay(display));
	close(evfd);
	exit(0);
}


void load_font(XFontStruct **font_info)
{
	char *fontname = FONT_NAME;

	if ((*font_info = XLoadQueryFont(display, fontname)) == NULL) {
		printf("Cannot open %s font\n", FONT_NAME);
		exit(1);
	}
}


void draw_point(int x, int y, int width, int size, unsigned long color)
{
	XSetForeground(display, gc, color);
	XSetLineAttributes(display, gc, width, LineSolid,
			   CapRound, JoinRound);
	XDrawLine(display, win, gc, x - size, y, x + size, y);
	XDrawLine(display, win, gc, x, y - size, x, y + size);
}


void point_blink(unsigned long color)
{
	int i, j;
	int cx, cy;
	static int shift = 0;

	if (points_touched != 4) {
		int RectDist = width / 200;
		i = points_touched / 2;
		j = points_touched % 2;
		cx = (MARK_POINT[j] * width) / SCREEN_MAX;
		cy = (MARK_POINT[i] * height) / SCREEN_MAX;

		XSetLineAttributes(display, gc, 1, LineSolid, CapRound, JoinRound);
		for (i = 0; i < NumRect; i++) {
			if ((i + shift) % NumRect == 0)
				XSetForeground(display, gc, BlinkCrossBG);
			else
				XSetForeground(display, gc, BlinkCrossFG);

#ifdef ROUND_SYMBOL
			XDrawArc(display, win, gc,
			         cx - i * RectDist, cy - i * RectDist,
			         i * (2 * RectDist), i * (2 * RectDist),
			         0, 359 * 64);
#else
			XDrawRectangle(display, win, gc,
				       cx - i * RectDist, cy - i * RectDist,
				       i * (2 * RectDist), i * (2 * RectDist));
#endif
		}
		shift++;
	}
}


void draw_message(char *msg)
{
	char buf[300];
	char *prompt[] = { buf };
#define num	(sizeof(prompt) / sizeof(prompt[0]))
	static int init = 0;
	static int p_len[num];
	static int p_width[num];
	static int p_height;
	static int p_maxwidth = 0;
	int i, x, y;
	int line_height;

	strncpy(buf, msg, sizeof buf);

	for (i = 0; i < num; i++) {
		p_len[i] = strlen(prompt[i]);
		p_width[i] = XTextWidth(font_info, prompt[i], p_len[i]);

		if (p_width[i] > p_maxwidth)
			p_maxwidth = p_width[i];
	}

	p_height = font_info->ascent + font_info->descent;
	init = 1;

	line_height = p_height + 5;
	x = (width - p_maxwidth) / 2;
	y = height / 2 - line_height;

	XSetForeground(display, gc, PromptText);
	XSetLineAttributes(display, gc, 3, LineSolid, CapRound, JoinRound);
	XClearArea(display, win, x - 8, y - 8 - p_height, p_maxwidth + 8 * 2,
	           num * line_height + 8 * 2, False);
	XDrawRectangle(display, win, gc, x - 8, y - 8 - p_height,
	               p_maxwidth + 8 * 2, num * line_height + 8 * 2);

	for (i = 0; i < num; i++) {
		XDrawString(display, win, gc, x, y + i * line_height, prompt[i],
			    p_len[i]);
	}
#undef num
}


void draw_text()
{
	static char *prompt[] = {
		"                    4-Pt Calibration",
		"Please touch the blinking symbol until beep or stop blinking",
		"                     (ESC to Abort)",
	};
#define num	(sizeof(prompt) / sizeof(prompt[0]))
	static int init = 0;
	static int p_len[num];
	static int p_width[num];
	static int p_height;
	static int p_maxwidth = 0;
	int i, x, y;
	int line_height;

	if (!init) {
		for (i = 0; i < num; i++) {
			p_len[i] = strlen(prompt[i]);
			p_width[i] = XTextWidth(font_info, prompt[i], p_len[i]);
			if (p_width[i] > p_maxwidth)
				p_maxwidth = p_width[i];
		}
		p_height = font_info->ascent + font_info->descent;
		init = 1;
	}
	line_height = p_height + 5;
	x = (width - p_maxwidth) / 2;
	y = height / 2 - 6 * line_height;

	XSetForeground(display, gc, PromptText);
	XClearArea(display, win, x - 11, y - 8 - p_height,
		   p_maxwidth + 11 * 2, num * line_height + 8 * 2, False);
	XSetLineAttributes(display, gc, 3, FillSolid,
			   CapRound, JoinRound);
	XDrawRectangle(display, win, gc, x - 11, y - 8 - p_height,
		       p_maxwidth + 11 * 2, num * line_height + 8 * 2);

	for (i = 0; i < num; i++) {
		XDrawString(display, win, gc, x, y + i * line_height, prompt[i],
			    p_len[i]);
	}
#undef num
}


void draw_graphics()
{
	int i, j;
	unsigned cx, cy;
	unsigned long color;

	draw_text();

	for (i = 0; i < 2; i++) {
		for (j = 0; j < 2; j++) {
			int num = 2 * i + j;

			if (num == points_touched)
				continue;
			
			if (num > points_touched)
				color = nTouchedCross;
			else
				color = TouchedCross;

			cx = (MARK_POINT[j] * width) / SCREEN_MAX;
			cy = (MARK_POINT[i] * height) / SCREEN_MAX;
			draw_point(cx, cy, width / 200, width / 64, color);
		}
	}
}


void get_gc(Window win, GC *gc, XFontStruct *font_info)
{
	unsigned long valuemask = 0;	/* ignore XGCvalues and use defaults */
	XGCValues values;
	unsigned int line_width = 5;
	int line_style = LineSolid;
	int cap_style = CapRound;
	int join_style = JoinRound;

	*gc = XCreateGC(display, win, valuemask, &values);

	XSetFont(display, *gc, font_info->fid);

	XSetLineAttributes(display, *gc, line_width, line_style,
			   cap_style, join_style);
}


int get_color()
{
	int default_depth;
	Colormap default_cmap;
	XColor my_color;
	int i;

	default_depth = DefaultDepth(display, screen);
	default_cmap = DefaultColormap(display, screen);

	for (i = 0; i < N_Colors; i++) {
		XParseColor(display, default_cmap, colors[i], &my_color);
		XAllocColor(display, default_cmap, &my_color);
		pixels[i] = my_color.pixel;
	}

	return 0;
}


Cursor create_empty_cursor()
{
	char nothing[] = { 0 };
	XColor nullcolor;
	Pixmap src = XCreateBitmapFromData(display, root, nothing, 1, 1);
	Pixmap msk = XCreateBitmapFromData(display, root, nothing, 1, 1);
	Cursor mycursor = XCreatePixmapCursor(display, src, msk,
					      &nullcolor, &nullcolor, 0, 0);
	XFreePixmap(display, src);
	XFreePixmap(display, msk);

	return mycursor;
}


void process_event()
{
	XEvent event;

	while (XCheckWindowEvent(display, win, -1, &event) == True) {
		switch (event.type) {
		case KeyPress:
			{
				KeySym keysym = XKeycodeToKeysym(display,
							     event.xkey.keycode, 0);

				if (keysym == XK_Escape) {
					puts("Aborted");
					cleanup_exit();
				}
			}
			break;

		case Expose:
			draw_graphics(win, gc, width, height);
			break;

		default:
			break;
		}
	}
}


double idle_time = 0;
double tick = 0;

void set_timer(double interval /* in second */ )
{
	struct itimerval timer;
	long sec = interval;
	long usec = (interval - sec) * 1.0e6;

	timer.it_value.tv_sec = sec;
	timer.it_value.tv_usec = usec;
	timer.it_interval = timer.it_value;
	setitimer(ITIMER_REAL, &timer, NULL);
	tick = interval;
}


void update_timer(void)
{
	int current = width * idle_time / IDLETIMEOUT;

	XSetLineAttributes(display, gc, 2, LineSolid, CapRound, JoinRound);
	XSetForeground(display, gc, Background);
	XDrawLine(display, win, gc, 0, height - 1, current, height - 1);
	XSetForeground(display, gc, TimerLine);
	XDrawLine(display, win, gc, current, height - 1, width, height - 1);
}


int register_fasync(int fd, void (*handle) (int))
{
	signal(SIGIO, handle);
	fcntl(fd, F_SETOWN, getpid());
	fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | FASYNC);
	return 0;
}


void sig_handler(int num)
{
	char buf[100];
	int i, rval, x, y;
	static int is_busy = 0;

	if (is_busy)
		return;
	is_busy = 1;

	switch (num) {
	case SIGALRM:
		if (!job_done)
			point_blink(BlinkCrossFG);

		update_timer();
		if (idle_time >= IDLETIMEOUT)
			cleanup_exit();

		idle_time += tick;
		XFlush(display);
		process_event();

		break;

	case SIGIO:
		rval = get_events(&x, &y);

		if (rval == -1)
			break;

		idle_time = 0;

		points_x[points_touched] = x;
		points_y[points_touched] = y;
		
		points_touched++;
		draw_graphics();

		break;

	default:
		break;
	}

	/* do the math */
	if (points_touched == 4 && !job_done) {
		int x_low, y_low, x_hi, y_hi;
		int x_min, y_min, x_max, y_max;
		int x_seg, y_seg;
		int x_inv = 0, y_inv = 0;

		/* get the averages */
		x_low = (points_x[0] + points_x[2]) / 2;
		y_low = (points_y[0] + points_y[1]) / 2;
		x_hi = (points_x[1] + points_x[3]) / 2;
		y_hi = (points_y[2] + points_y[3]) / 2;

		/* see if one of the axes is inverted */
		if (x_low > x_hi) {
			int tmp = x_hi;
			x_hi = x_low;
			x_low = tmp;
			x_inv = 1;
		}
		if (y_low > y_hi) {
			int tmp = y_hi;
			y_hi = y_low;
			y_low = tmp;
			y_inv = 1;
		}


		/* calc the min and max values */
		x_seg = (x_hi - x_low) / (SCREEN_DIVIDE - 2);
		x_min = x_low - x_seg;
		x_max = x_hi + x_seg;

		y_seg = (y_hi - y_low) / (SCREEN_DIVIDE - 2);
		y_min = y_low - y_seg;
		y_max = y_hi + y_seg;

		/* print it, hint: evtouch has Y inverted */
		printf("Copy-Paste friendly, for evtouch XFree86 driver\n");
		printf("	Option \"MinX\" \"%d\"\n", x_inv ? x_max : x_min);
		printf("	Option \"MinY\" \"%d\"\n", y_inv ? y_min : y_max);
		printf("	Option \"MaxX\" \"%d\"\n", x_inv ? x_min : x_max);
		printf("	Option \"MaxY\" \"%d\"\n", y_inv ? y_max : y_min);

		draw_message("   Done...   ");
		XFlush(display);

		job_done = 1;
		idle_time = IDLETIMEOUT * 3 / 4;
		update_timer();
	}

	is_busy = 0;

	return;
}


int main(int argc, char *argv[], char *env[])
{
	char *display_name = NULL;
	XSetWindowAttributes xswa;

	/* one arg: device name */
	if (argc != 2) {
		fprintf(stderr, "Usage %s <device>!\n", argv[0]);
		return 1;
	}

	evfd = open(argv[1], O_RDONLY | O_NONBLOCK);
	if (evfd == -1) {
		fprintf(stderr, "Cannot open device file!\n");
		return 1;
	}


	/* connect to X server */
	if ((display = XOpenDisplay(display_name)) == NULL) {
		fprintf(stderr, "%s: cannot connect to X server %s\n",
			progname, XDisplayName(display_name));
		close(evfd);
		exit(1);
	}

	screen = DefaultScreen(display);
	root = RootWindow(display, screen);

	/* setup window attributes */
	xswa.override_redirect = True;
	xswa.background_pixel = BlackPixel(display, screen);
	xswa.event_mask = ExposureMask | KeyPressMask;
	xswa.cursor = create_empty_cursor();

	/* get screen size from display structure macro */
	width = DisplayWidth(display, screen);
	height = DisplayHeight(display, screen);

	win = XCreateWindow(display, RootWindow(display, screen),
	                    0, 0, width, height, 0,
	                    CopyFromParent, InputOutput, CopyFromParent,
	                    CWOverrideRedirect | CWBackPixel | CWEventMask |
	                    CWCursor, &xswa);
	XMapWindow(display, win);
	XGrabKeyboard(display, win, False, GrabModeAsync, GrabModeAsync,
	              CurrentTime);
	XGrabServer(display);
	load_font(&font_info);
	get_gc(win, &gc, font_info);
	get_color();

	XSetWindowBackground(display, win, Background);
	XClearWindow(display, win);

	signal(SIGALRM, sig_handler);
	set_timer(BLINKPERD);
	register_fasync(evfd, sig_handler);

	/* wait for signals */
	while (1)
		pause();

	return 0;
}


evtouch.h

/*
 *
 * Copyright 2004 by Kenan Esau <kenan.esau@conan.de>, Baltmannsweiler, 
 * Germany.
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the names of copyright holders not be
 * used in advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.  The copyright holders
 * make no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without express or
 * implied warranty.
 *
 * THE COPYRIGHT HOLDERS DISCLAIM ALL WARRANTIES WITH REGARD TO THIS
 * SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS, IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 */


#ifndef _evtouch_H_
#define _evtouch_H_


/******************************************************************************
 *  Definitions
 *  structs, typedefs, #defines, enums
 *****************************************************************************/

/* Physical Screen Dimensions. (for default values)
   For the Lifebook B Series currently 800x600 pixels */
#define EV_SCREEN_WIDTH       800
#define EV_SCREEN_HEIGHT      600
#define EV_AXIS_MIN_RES       0
#define EV_AXIS_MAX_RES       1024
#define EV_PAN_BORDER         12

#define EV_TIMEOUT            500

/*
 * The event structure itself
 */


/*
 * Event types
 */
#define EV_SYN			0x00
#define EV_KEY			0x01
#define EV_REL			0x02
#define EV_ABS			0x03

/*
 * Absolute axes
 */
#define ABS_X			0x00
#define ABS_Y			0x01

/*
 * Buttons
 */
#define BTN_LEFT		0x110
#define BTN_RIGHT		0x111
#define BTN_TOUCH		0x14a

#define TOUCHED 0x01
#define X_COORD 0x02
#define Y_COORD 0x04
#define LB_STAT 0x08  /* LB up / down */
#define RB_STAT 0x10  /* RB up / down (both needed for 3-btn emu) */

typedef struct _EVTouchPrivateRec
{
        int min_x;  /* Minimum x reported by calibration        */
        int max_x;  /* Maximum x                    */
        int min_y;  /* Minimum y reported by calibration        */
        int max_y;  /* Maximum y                    */

        int click_timer;
        int drag_timer;
        int emulate3;
        int emulate3_timeout;
        Bool emulate3_timer_expired;

        int move_limit;

        int cur_x;
        int cur_y;
        int old_x;
        int old_y;
        /* pointers to the current viewport coordinates */
        int *pViewPort_X0;    /* Min X */
        int *pViewPort_X1;    /* Max X */
        int *pViewPort_Y0;    /* Min Y */
        int *pViewPort_Y1;    /* Max Y */
        int virtual;          /* virtual=1 indicates that there is a virtual screen */
        int x;                /* x in screen coords */
        int y;                /* y in screen coords */
        int phys_width;       /* Physical X-Resolution */
        int phys_height;      /* Physical Y-Resolution */

        struct input_event ev_touched;
        int touch_x;
        int touch_y;
        unsigned char touch_flags; /* 1 - touched, 2 - x-coord received
                                      4 - y-coord received */
        Bool drag;
        Bool currently_dragging;
        Bool pan_viewport;

        int button_number;   /* which button to report */
        int reporting_mode;   /* TS_Raw or TS_Scaled */

        int screen_num;    /* Screen associated with the device */
        int screen_width;   /* Width of the associated X screen  */
        int screen_height;   /* Height of the screen              */
        
        XISBuffer *buffer;
        struct input_event ev; /* packet being/just read */
        
        int packeti;    /* index into packet */
        Bool cs7flag;
        Bool binary_pkt;   /* indicates packet was a binary touch */

        int bin_byte;    /* bytes recieved in binary packet */

        LocalDevicePtr local;
} EVTouchPrivateRec, *EVTouchPrivatePtr;

/******************************************************************************
 *  Declarations
 *****************************************************************************/
/*int DumpOpts (XF86OptionPtr opts); */
static Bool DeviceControl (DeviceIntPtr dev, int mode);
static Bool DeviceOn (DeviceIntPtr dev);
static Bool DeviceOff (DeviceIntPtr dev);
static Bool DeviceInit (DeviceIntPtr dev);
static void ReadInput (LocalDevicePtr local);
/* static int ControlProc (LocalDevicePtr local, xDeviceCtl * control); */
static void ControlProc(DeviceIntPtr device, PtrCtrl *ctrl);
static void CloseProc (LocalDevicePtr local);
static int SwitchMode (ClientPtr client, DeviceIntPtr dev, int mode);
static Bool ConvertProc (LocalDevicePtr local, int first, int num, int v0, int v1, int v2, int v3, int v4, int v5, int *x, int *y);
static Bool QueryHardware (LocalDevicePtr local);
static void EVTouchNewPacket (EVTouchPrivatePtr priv);

/*static Bool EVTouchSendCommand(LocalDevicePtr local, unsigned char command[]);*/
static unsigned char EVTouchRead (EVTouchPrivatePtr priv);
static Bool EVTouchGetPacket (EVTouchPrivatePtr priv);
void EVTouchLBRBEvent(EVTouchPrivatePtr priv);

#ifdef LOG_RAW_PACKET
#define EVTouchDumpPacketToLog(priv) ( xf86ErrorFVerb(2, "EVTOUCHPCKT %d %d %d\n",                                       priv->ev.type, priv->ev.code, priv->ev.value ) )
#else
#define EVTouchDumpPacketToLog(priv)
#endif
/* 
 *    DO NOT PUT ANYTHING AFTER THIS ENDIF
 */
#endif
